presto-benchmark-runner/src/main/java/com/facebook/presto/benchmark/executor/ConcurrentPhaseExecutor.java

```diff
@@ -0,0 +1,120 @@
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorCompletionService;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static com.facebook.presto.benchmark.event.BenchmarkQueryEvent.EventStatus;
+import static com.facebook.presto.benchmark.event.BenchmarkQueryEvent.EventStatus.FAILED;
+import static com.facebook.presto.benchmark.event.BenchmarkQueryEvent.EventStatus.SUCCEEDED;
+import static java.lang.Thread.currentThread;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newFixedThreadPool;
+
+public class ConcurrentPhaseExecutor
+        implements PhaseExecutor
+{
+    private static final Logger log = Logger.get(ConcurrentPhaseExecutor.class);
+
+    private final int maxConcurrency;
+
+    private BenchmarkQueryExecutor benchmarkQueryExecutor;
+    private List<String> queries;
+    private Map<String, BenchmarkQuery> queryMap;
+    private Map<String, String> sessionProperties;
+
+    private ExecutorService executor;
+    private CompletionService<BenchmarkQueryEvent> completionService;
+    private AtomicInteger queriesSubmitted = new AtomicInteger();
+
+    public ConcurrentPhaseExecutor(
+            BenchmarkRunnerConfig config,
+            BenchmarkQueryExecutor benchmarkQueryExecutor,
+            List<String> queries,
+            Map<String, BenchmarkQuery> queryMap,
+            Map<String, String> sessionProperties)
+    {
+        this.maxConcurrency = requireNonNull(config, "maxConcurrency is null").getMaxConcurrency();
+        this.benchmarkQueryExecutor = requireNonNull(benchmarkQueryExecutor, "benchmarkQueryExecutor is null");
+        this.queries = ImmutableList.copyOf(requireNonNull(queries, "queries is null"));
+        this.queryMap = ImmutableMap.copyOf(queryMap);
+        this.sessionProperties = ImmutableMap.copyOf(sessionProperties);
+    }
+
+    @Override
+    public void run()
+    {
+        this.executor = newFixedThreadPool(maxConcurrency);
+        this.completionService = new ExecutorCompletionService<>(executor);
+
+        for (String query : queries) {
+            completionService.submit(() -> benchmarkQueryExecutor.run(queryMap.get(query), sessionProperties));
+        }
+
+        this.queriesSubmitted.addAndGet(queries.size());
+        reportProgressUntilFinished();
+    }
+
+    private void reportProgressUntilFinished()
+    {
+        int completed = 0;
+        double lastProgress = 0;
+        Map<EventStatus, Integer> statusCount = new EnumMap<>(EventStatus.class);
+
+        while (completed < queriesSubmitted.get()) {
+            try {
+                BenchmarkQueryEvent event = completionService.take().get();
+                completed++;
+                statusCount.compute(EventStatus.valueOf(event.getStatus()), (status, count) -> count == null ? 1 : count + 1);
+                if (event.getStatus().equals(FAILED.name())) {
```

Comment:
Let's provide a method that return `QueryEventStatus` instead of String. You may call it `getEventStatus`